### ISSUE TITLE
when to use early-stopping? in cv stage or training stage?

### ISSUE THREAD

**Issue by LancelotHolmes** [dmlc/xgboost#2261]
I'm confused about when to use the early_stopping, say if my pipeline is like:

* k-fold cross validation to tune the model params 
* use all training data to train the model 
* finally predict on the test set

my question is when should we use early_stopping, cv stage or training stage?

if I use the early_stopping in cv stage, then how can I choose the best_ntree_limit or best iteration on my training stage, or should I separate my training set to use like 20%  training data to eval and get the best_ntree_limit which can be used on the prediction? 

for example:pipeline 1,use early stopping on train stage but we need to split the train set to get an eval set

* first use cv to select model or tune model params
* then split the train set like(80%-20%) and use early-stopping to train/fit the model and get the best_ntree_limit 
* finally use the model(estimator) return from train stage to predict on the test set with the best_ntree_limit 

pipeline 2,use early stopping in cv stage, according to this [post](http://stackoverflow.com/a/40577965/6653189)

* k-fold cv to tune params and also get the best iteration
* count the train iterations according to the cv folds and cv iteration by 
> best_nrounds = int(cv_best_iterations / (1 - 1 / n_folds))
and train the data directly by the best_nrounds counted
* finally predict on the test set with the model(estimator) after training

which pipeline is correct? Or what's the correct pipeline?

**Comment by mik1904** [comment]
@LancelotHolmes Have you had any answer in this? I'm facing the same problem.
Thank you

**Comment by LancelotHolmes** [comment]
In my competition,I try both methods and find that use cv to tune my model and find the best params groups, and then train by 80% of whole training data for the final estimator and  split 20% of whole training data to evaluate my estimator to get control the boosting rounds and finally use the finla estimator to predict on the test data with `bst_ntree_limit` estimator.

While use cv to count the iteration rounds performs bad,so I just use cv to get the params and use early stopping to avoid over-fitting.
@mik1904
